### PR TITLE
Fix: Add skipContextCompression flag to prevent unwanted compression - Fixes #4430

### LINKED ISSUE #4430: [BUG] Context condensing triggers on settings Save during active task
## Type
Bug

## Problem
Saving settings during an active task immediately triggers context compression, even when the configured threshold is 100%.

## Context
This affects users running long, read-heavy tasks. When the settings view is opened and Save is clicked during an active task, a context compression event occurs right away and the context shrinks, despite the context bar being well below the threshold. Environment example: GPT‑4.1 via OpenRouter with “Multi-file read (experimental)” enabled.

## Reproduction
1) Set your model to GPT‑4.1 with OpenRouter provider
2) Enable “Multi-file read (experimental)” in settings
3) Set context compression threshold to 100%
4) Start a new task in a workspace with a good amount of code
5) Enter the prompt: `read 10 files at a time 8 times in a row`
6) While the task is running and the context bar is increasing, open the settings view
7) Toggle any setting back and forth to activate the Save button (without actually changing any setting)
8) Click Save
9) Observe that context compression triggers immediately, regardless of actual context usage

## Expected result
Context compression should only trigger when the configured threshold is reached (100% in this case). Saving settings during an active task should not cause compression by itself.

## Actual result
Context compression triggers immediately upon saving settings during an active task, even when context usage is well below the threshold.

### PR DESCRIPTION
**Pull request by shirishgone**
…- Fixes #4430

This PR implements a `skipContextCompression` flag to prevent unwanted context compression during multi-file read operations and settings save operations. The fix addresses the issue where context compression was being triggered inappropriately, causing performance issues and unexpected behavior.

**Key implementation details:**
- Added `skipContextCompression` parameter to `truncateConversationIfNeeded` function
- When the flag is `true`, the function returns early without performing compression, even if the context usage exceeds the threshold
- The flag is designed to be used in scenarios where compression should be temporarily disabled (e.g., during multi-file reads, settings operations)
- Maintains backward compatibility by making the parameter optional with a default value of `false`

**Reviewers should pay attention to:**
- The early return logic in the sliding window function
- Test coverage for both skip and normal compression scenarios
- Integration points where the flag needs to be passed through

### Test Procedure

**How I tested this implementation:**
1. Created comprehensive unit tests covering:
   - Skip functionality when flag is `true`
   - Normal compression when flag is `false`
   - Edge cases with different threshold values
   - Multi-file read scenario simulation

2. All existing tests continue to pass, ensuring no regression

**How reviewers can verify:**
1. Run the test suite: `pnpm test`
2. Check the new test file: `src/core/sliding-window/__tests__/context-compression-fix.test.ts`
3. Verify that the `skipContextCompression` parameter is properly handled in the main function

**Testing environment:** 
- Node.js 20.19.2
- All tests passing with vitest framework

### Pre-Submission Checklist

- [x] **Issue Linked**: This PR is linked to an approved GitHub Issue (#4430)
- [x] **Scope**: My changes are focused on the linked issue (context compression skip functionality)
- [x] **Self-Review**: I have performed a thorough self-review of my code
- [x] **Testing**: New comprehensive tests have been added to cover the skip functionality
- [x] **Documentation Impact**: No user-facing documentation updates required (internal API change)
- [x] **Contribution Guidelines**: I have read and agree to the Contributor Guidelines

### Documentation Updates

- [x] No documentation updates are required (internal API enhancement)

### Additional Notes

This fix specifically addresses the performance issue mentioned in #4430 where context compression was being triggered during multi-file read operations when `maxConcurrentFileReads > 1`. The `skipContextCompression` flag provides a clean way to temporarily disable compression during these operations while maintaining the existing compression logic for normal use cases.

The implementation is designed to be:
- **Safe**: Backward compatible with existing code
- **Targeted**: Only affects compression behavior when explicitly requested
- **Testable**: Comprehensive test coverage for all scenarios

### Get in Touch
shirish.gone@smartsheet.com

----

> [!IMPORTANT]
> Introduces `skipContextCompression` flag to control context compression, preventing unwanted compression during multi-file reads and settings operations.
> 
>   - **Behavior**:
>     - Adds `skipContextCompression` flag to `truncateConversationIfNeeded` in `index.ts` to prevent unwanted context compression.
>     - When `true`, skips compression even if context usage exceeds threshold.
>     - Used during multi-file reads and settings operations to avoid performance issues.
>     - Backward compatible with default `false` value.
>   - **Integration**:
>     - `Task.ts`: Passes `skipContextCompression` flag in `truncateConversationIfNeeded` calls.
>     - `webviewMessageHandler.ts`: Sets `_skipNextContextCompressionCheck` when `maxConcurrentFileReads > 1`.
>   - **Testing**:
>     - New tests in `context-compression-fix.test.ts` for scenarios with and without the flag.
>     - Tests cover edge cases and multi-file read scenarios.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for eb95b87f8c4e959926cf98b67a0c29c8646fd7bb. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: src/core/sliding-window/index.ts [hunk 1]
```diff
@@ -77,6 +77,10 @@ type TruncateOptions = {
 	condensingApiHandler?: ApiHandler
 	profileThresholds: Record<string, number>
 	currentProfileId: string
+	// Context for batch operations and settings saves - cleaner architecture
+	isBatchActive?: boolean
+	isSettingsSave?: boolean
+	maxConcurrentFileReads?: number
```
Comment (ghost) [inline src/core/sliding-window/index.ts]: These properties (, , ) are added to the type but never actually used in the implementation. Could we remove these unused properties to keep the interface clean?

File: src/core/sliding-window/index.ts [hunk 2]
```diff
@@ -102,6 +106,7 @@
 	condensingApiHandler,
 	profileThresholds,
 	currentProfileId,
+	skipContextCompression,
```
Comment (ghost) [inline src/core/sliding-window/index.ts:106]: Is  missing from the destructured parameters? The parameter is used on line 153 but not extracted from the options object here.

File: src/core/webview/webviewMessageHandler.ts [hunk 3]
```diff
@@ -1288,6 +1288,10 @@ export const webviewMessageHandler = async (
 		case "maxConcurrentFileReads":
 			const valueToSave = message.value // Capture the value intended for saving
 			await updateGlobalState("maxConcurrentFileReads", valueToSave)
+			const activeTask = provider.getCurrentCline()
+			if (activeTask && typeof valueToSave === "number" && valueToSave > 1) {
+				;(activeTask as any)._skipNextContextCompressionCheck = true
```
Comment (ghost) [inline src/core/webview/webviewMessageHandler.ts]: The fix only handles the  setting, but the issue mentions "settings save operations" in general. Should we also handle other settings that might trigger unwanted compression during save operations?

**Comment by daniel-lxs** [comment]
Hey @shirishgone, Thank you for your contribution. I think we should try and address the root cause of the bug rather than trying to mitigate the symptoms. The main concern is that this will prevent condensing when it's actually needed if the files read exceed the context window of the model.

I'll close this PR for now, but feel free to continue the discussion. I would like to know what you think.
